hazelcast/src/test/java/com/hazelcast/test/AbstractHazelcastClassRunner.java

```diff
@@ -183,6 +187,53 @@ private Repeat getRepeatable(FrameworkMethod method) {
     }
 
     @Override
+    protected List<TestRule> getTestRules(Object target) {
+        List<TestRule> testRules = super.getTestRules(target);
+
+        Set<Class<? extends TestRule>> testRuleClasses = new HashSet<Class<? extends TestRule>>();
+
+        TestClass testClass = getTestClass();
+
+        // Find the required test rule classes from test class itself
+        Annotation[] classAnnotations = testClass.getAnnotations();
+        for (Annotation annotation : classAnnotations) {
+            Class<? extends Annotation> annotationType = annotation.annotationType();
+            AutoRegisteredTestRule autoFilterRule = annotationType.getAnnotation(AutoRegisteredTestRule.class);
+            if (autoFilterRule != null) {
+                Class<? extends TestRule> testRuleClass = autoFilterRule.testRule();
+                testRuleClasses.add(testRuleClass);
+            }
+        }
+
+        // Find the required test rule classes from methods
+        List<FrameworkMethod> annotatedMethods = testClass.getAnnotatedMethods();
+        for (FrameworkMethod annotatedMethod : annotatedMethods) {
+            Annotation[] methodAnnotations = annotatedMethod.getAnnotations();
+            for (Annotation annotation : methodAnnotations) {
```

Comment:
i couldn't get the idea here, 
first question:  this `getTestRules` is called for each methods or for each classes?
second question: if it calls for each class, you collect all rules and apply all methods?
Can you clarify please?